### PR TITLE
Implement advanced verification of SD-JWT in Keycloak

### LINKED ISSUE #30907: Implement advanced verification of SD-JWT in Keycloak
### Description

The current native implementation of SD-JWT in keycloak is focussed toward producing SD-JWT. This was necessary for the implementation of credential issuance:

- #25936 

Further, keycloak shall be given the capability of verifying an SD-JWT, asit does with verifying token

- #29859 

In that context following verification have to be implemented to comply with the SD-JWT specification.

https://drafts.oauth.net/oauth-selective-disclosure-jwt/draft-ietf-oauth-selective-disclosure-jwt.html#name-verification-of-the-sd-jwt
https://drafts.oauth.net/oauth-selective-disclosure-jwt/draft-ietf-oauth-selective-disclosure-jwt.html#name-verification-by-the-verifie
https://drafts.oauth.net/oauth-selective-disclosure-jwt/draft-ietf-oauth-selective-disclosure-jwt.html#name-decoy-digests-2

### Discussion

https://github.com/keycloak/keycloak/discussions/29857

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by IngridPuppet**
Closes #30907

### CODE REVIEW

**Comment by mposolda** [comment]
@wistefan @francis-pouatcha @tnorimat Anyone of you available to review this please?

**Comment by tnorimat** [comment]
@mposolda Some test failed, but it seems that these failed tests are not relevant to the PR. Could you re-run the CI for the PR?

**Comment by tnorimat** [comment]
@mposolda The re-ran CI also had one fail, but the fail is different from the one caused by the previous-ran CI.
Maybe, the fail was not caused by the PR.

**Comment by ahus1** [comment]
The CI rerun for the job failed as it was missing the archive which was expired. I rerun the full CI. Let's see if that succeeds.

**Comment by ahus1** [comment]
@wistefan @francis-pouatcha @tnorimat - the tests are now green. Please continue your review when you have the time.

**Comment by tnorimat** [comment]
@ahus1 Thank you. I will review the PR.

File: core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+        return KeyBindingJwtVerificationOpts.builder()
+                .withKeyBindingRequired(true)
+                .withAllowedMaxAge(Integer.MAX_VALUE)
+                .withNonce("1234567890")
+                .withAud("https://verifier.example.org")
+                .withValidateExpirationClaim(false)
+                .withValidateNotBeforeClaim(false);
+    }
+
+    private ObjectNode exampleS20KbPayload() {
```
Comment (tnorimat) [inline core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:415]: What does `S20` means? If possible, could you name the method that other people can understand?

**Review by tnorimat — CHANGES_REQUESTED** [review]
@IngridPuppet Thank you for the PR. I added some review comments. Could you check them?

File: core/src/main/java/org/keycloak/sdjwt/SdJws.java [hunk 2]
```diff
@@ -107,4 +107,65 @@ private static final JsonNode readPayload(JWSInput jwsInput) {
...
+    /**
+     * Verifies that SD-JWT was issued by one of the provided issuers. Verification is case-insensitive
+     * @param issuers List of trusted issuers
+     */
+    public void verifyIssClaim(List<String> issuers) throws VerificationException {
+        verifyClaimAgainstTrustedValues(issuers, "iss");
+    }
+
+    /**
+     * Verifies that SD-JWT vct claim matches the expected one. Verification is case-insensitive
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJws.java:151]: I've removed that relaxation. It is now up to the caller to possibly match case-insensitively, according to its policy.

File: core/src/test/java/org/keycloak/sdjwt/SdJwtVerificationTest.java [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+    }
+
+    private SdJwt.Builder exampleFlatSdJwtV2(ObjectNode claimSet, DisclosureSpec disclosureSpec) {
+        return SdJwt.builder()
+                .withDisclosureSpec(disclosureSpec)
+                .withClaimSet(claimSet)
+                .withSigner(testSettings.issuerSigContext);
+    }
+
+    private SdJwt.Builder exampleSdJwtWithUndisclosedNestedFieldsV1() {
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/test/java/org/keycloak/sdjwt/SdJwtVerificationTest.java:322]: I acknowledge that the methods are in fact largely similar, but they are not the same. Method `exampleRecursiveSdJwtV1()` has this line:

```
.withUndisclosedClaim("address", "BZFzhQsdPfZY1WSL-1GXKg")
```

By making the whole address object selectively disclosable, we can test that the implemented logic can recursively process the object accordingly.

To address this point, I've refactored the methods to decrease code duplication and added a comment above the key line addition.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+            JWK jwk = SdJwtUtils.mapper.convertValue(cnfJwk, JWK.class);
+            keyWrapper = JWKSUtils.getKeyWrapper(jwk);
+            Objects.requireNonNull(keyWrapper);
+        } catch (Exception e) {
+            throw new VerificationException("Malformed or unsupported cnf/jwk claim");
+        }
+
+        // Build verifier
+        if (keyWrapper.getType().equals(KeyType.EC)) {
+            if (keyWrapper.getAlgorithm() == null) {
```
[the remark this answers is not in the document]
Comment (tnorimat) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:253]: Sorry, it is my misunderstanding.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+     *
+     * <p>
+     * By recursively disclosing all disclosable fields in the SdJwt payload, validation rules are
+     * enforced regarding the conformance of linked disclosures. Additional rules should be enforced
+     * after calling this method based on the visited data arguments.
+     * </p>
+     *
+     * @return the fully disclosed SdJwt payload
+     */
+    private JsonNode validateViaRecursiveDisclosing(
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:400]: I agree that enforcing this rule should also be implemented. I have added the code for it.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+                        break;
+                    case "P-521":
+                        alg = "ES512";
+                        break;
+                }
+
+                keyWrapper.setAlgorithm(alg);
+            }
+
+            return new ECDSASignatureVerifierContext(keyWrapper);
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:271]: Sorry, I made a mistake while searching for the available implementations of the `SignatureVerifierContext` interface in the codebase. I've now added support for RSA key binding signatures.

As for EdDSA, while there is a corresponding `SignatureVerifierContext` implementation, I am afraid there isn't yet support for parsing OKP JWKs into a `KeyWrapper`, a feature on which the current logic relies.

https://github.com/keycloak/keycloak/blob/073a6373628489b65e0d192e075957d58ed371ea/core/src/main/java/org/keycloak/jose/jwk/AbstractJWKParser.java#L114-L116

I additionally think that this PR is not the right place to bridge this gap. I would however welcome any further guidance.

File: core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java (hunk 1 above)
Comment (IngridPuppet) [reply core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:415]: I've removed the `S20` particle to make it less obscure. The method has no further intent beyond pre-constructing a key binding jwt payload.

File: core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+                )
+        );
+
+        assertEquals(exceptionMessage, exception.getMessage());
+        if (exceptionCauseMessage != null) {
+            assertThat(exception.getCause().getMessage(), containsString(exceptionCauseMessage));
+        }
+    }
+
+    private void testShouldFailGenericS20(
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:366]: I've renamed the method to `testShouldFailGeneric2` to make it less obscure, then added a comment to specify that the base sdjwt it considers is sample `sdjwt/s20.1-sdjwt+kb.txt`.

File: core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+
+    @ClassRule
+    public static CryptoInitRule cryptoInitRule = new CryptoInitRule();
+
+    static ObjectMapper mapper = new ObjectMapper();
+    static TestSettings testSettings = TestSettings.getInstance();
+
+    @Test
+    public void testVerif_s20_1_sdjwt_with_kb() throws VerificationException {
+        String sdJwtVPString = TestUtils.readFileAsString(getClass(), "sdjwt/s20.1-sdjwt+kb.txt");
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:56]: This point and the ones alike below are concerned about the origin of the following files used in the testsuite:

- s20.1-sdjwt+kb.txt
- s20.1-sdjwt+kb--wrong-kb-signature.txt
- s20.2-sdjwt+kb--no-cnf-claim.txt
- s20.3-sdjwt+kb--wrong-kb-typ.txt
- s20.6-sdjwt+kb--disclosure-with-no-digest.txt
- s20.7-sdjwt+kb--invalid-array-elt-disclosure.txt
- s20.7-sdjwt+kb--invalid-field-disclosure.txt
- s20.8-sdjwt+kb--cnf-hmac.txt
- s20.8-sdjwt+kb--cnf-is-not-jwk.txt
- s20.8-sdjwt+kb--cnf-jwk-is-malformed.txt
- s20.8-sdjwt+kb--cnf-rsa-ps256.txt
- s20.8-sdjwt+kb--cnf-rsa-ps512.txt
- s20.8-sdjwt+kb--cnf-rsa-rs256.txt
- s20.8-sdjwt+kb--es384.txt
- s20.8-sdjwt+kb--es512.txt

All of them were manually crafted. The first one pairs an sdjwt that was present in the project with a corresponding key binding JWT issued in another test. The other ones use online tools such as jwt.io to create alterations for alternative scenarios.

I have added a global comment to specify that. Is that appropriate?

**Review by tnorimat — COMMENTED** [review]
@IngridPuppet Thank you. Could you check review comments?

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+     * </p>
+     *
+     * @throws VerificationException if verification failed
+     */
+    private void preventKeyBindingJwtReplay(
+            KeyBindingJwtVerificationOpts keyBindingJwtVerificationOpts
+    ) throws VerificationException {
+        JsonNode nonce = keyBindingJwt.getPayload().get("nonce");
+        if (nonce == null || !nonce.isTextual()
+                || !nonce.asText().equals(keyBindingJwtVerificationOpts.getNonce())) {
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:664]: Do you mean that the wallet reads the `nonce` from the authorization request when creating the key binding JWT? Yes. Here is how we instantiate `keyBindingJwtVerificationOpts` in a project:

```java
var nonce = requestObjectDTO.getNonce();
var clientId = requestObjectDTO.getClientId();
        
var keyBindingJwtVerificationOpts = keyBindingJwtVerificationOptsBuilder
        .withKeyBindingRequired(isKeyBindingRequired)
        .withAud(clientId)
        .withNonce(nonce)
        .build();
```

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 10]
```diff
@@ -0,0 +1,709 @@
...
+            throw new VerificationException("Key binding JWT: Unexpected `nonce` value");
+        }
+
+        JsonNode aud = keyBindingJwt.getPayload().get("aud");
+        if (aud == null || !aud.isTextual()
+                || !aud.asText().equals(keyBindingJwtVerificationOpts.getAud())) {
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:670]: Do you mean that the wallet read the `aud` value from the authorization request (field `client_id`) when creating the key binding JWT? Yes. Here is how we instantiate `keyBindingJwtVerificationOpts` in a project:

```java
var nonce = requestObjectDTO.getNonce();
var clientId = requestObjectDTO.getClientId();
        
var keyBindingJwtVerificationOpts = keyBindingJwtVerificationOptsBuilder
        .withKeyBindingRequired(isKeyBindingRequired)
        .withAud(clientId)
        .withNonce(nonce)
        .build();
```

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 11]
```diff
@@ -0,0 +1,709 @@
...
+     * <p>
+     * If any Disclosure was not referenced by digest value in the Issuer-signed JWT (directly or recursively via
+     * other Disclosures), the SD-JWT MUST be rejected.
+     * </p>
+     *
+     * @throws VerificationException if not the case
+     */
+    private void validateDisclosuresVisits(Set<String> visitedDisclosureStrings)
+            throws VerificationException {
+        if (visitedDisclosureStrings.size() < disclosures.size()) {
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:644]: Yes, you're right: we want to check that all disclosures have been visited. So when the condition is NOT met, then we throw an exception.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 12]
```diff
@@ -0,0 +1,709 @@
...
+        // Check that the creation time of the Key Binding JWT, as determined by the iat claim,
+        // is within an acceptable window
+
+        try {
+            keyBindingJwt.verifyIssuedAtClaim();
+        } catch (VerificationException e) {
+            throw new VerificationException("Key binding JWT: Invalid `iat` claim", e);
+        }
+
+        long now = Instant.now().getEpochSecond();
```
[the remark this answers is not in the document]
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java]: Agree. To further promote re-usability, I've created an appropriate method in the `SdJws` superclass.

**Comment by IngridPuppet** [comment]
@tnorimat Upon re-running tests locally, I realized there was an issue with the `WildFlyElytronProvider` not matching RSASSA-PSS signatures with JavaAlgorithm's labels. In your review, please pay extra attention to the updates I made to make it work for SD-JWT verification.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:669]: @IngridPuppet Thank you for the clarification.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:663]: @IngridPuppet Thank you for the clarification.

**Review by tnorimat — APPROVED** [review]
LGTM

**Comment by tnorimat** [comment]
@mposolda @wistefan @francis-pouatcha If possible, could you also check the PR?

**Comment by mposolda** [comment]
Looks good to me, but has limited knowledge of oid4vc code

@francis-pouatcha Can you perhaps review as well please?

**Comment by francis-pouatcha** [comment]
@mposolda  I will review this PR tomorrow.

**Review by francis-pouatcha — APPROVED** [review]

**Comment by francis-pouatcha** [comment]
Looked all good for me. Major feedback already addressed by @tnorimat. Requested merged of current main branch.

**Comment by IngridPuppet** [comment]
Thank you for your feedback. Some updates:
- Rebased on main and fixed the change that was breaking the compilation, at least locally — @mposolda.
- Noticed that verification couldn't be repeated because state was changed. To address this, I've deep cloned a problematic argument in commit https://github.com/keycloak/keycloak/pull/30966/commits/2e4f38c792aaa37c10d4f00d410b1d889bcf1b60. Please check again — @francis-pouatcha @tnorimat.

**Comment by francis-pouatcha** [comment]
@IngridPuppet [2e4f38c](https://github.com/keycloak/keycloak/commit/2e4f38c792aaa37c10d4f00d410b1d889bcf1b60) looks good for me.

**Review by francis-pouatcha — APPROVED** [review]

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+
+        // Build verifier
+
+        // KeyType.EC
+        if (keyWrapper.getType().equals(KeyType.EC)) {
+            if (keyWrapper.getAlgorithm() == null) {
+                Objects.requireNonNull(keyWrapper.getCurve());
+
+                String alg = null;
+                switch (keyWrapper.getCurve()) {
```
Comment (tnorimat) [inline core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:260]: This switch clause can be represented simply as follows.
```
                String alg = switch (keyWrapper.getCurve()) {
                    case "P-256" -> "ES256";
                    case "P-384" -> "ES384";
                    case "P-521" -> "ES512";
                    default -> null;
                };
```

**Review by tnorimat — APPROVED** [review]
@IngridPuppet Thank you. I approved the PR. I added one review comment but it is up to you to accept my recommendation for coding in the comment.

File: core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java (hunk 13 above)
Comment (IngridPuppet) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:260]: Thank you for recommending this simpler syntax. Actually, my IDE has also been prompting for this. However, as the `core` module in locked in Java 11, we cannot use switch expressions.
https://github.com/keycloak/keycloak/blob/adb2af442a88b88452cd8aee2d064c906d4eb9ec/core/pom.xml#L34-L39

Comment (tnorimat) [reply core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:260]: Thank you for the clarification.

**Comment by tnorimat** [comment]
@mposolda Hello, I and @francis-pouatcha  approved the PR. Could you check the PR?

**Review by mposolda — APPROVED** [review]
@IngridPuppet @tnorimat @francis-pouatcha Thanks for all the updates and for the review of this PR!

I hope to merge if tests are OK

**Comment by mabartos** [comment]
@mposolda just one thing - might be good to mention it in the release notes, right?

**Pull request merged**